src/models/room.ts

```diff
@@ -940,12 +955,42 @@ export class Room extends ReadReceipt<RoomEmittedEvents, RoomEventHandlerMap> {
         if (nonFunctionalMemberCount > 2) return;
 
         // Prefer the list of heroes, if present. It should only include the single other user in the DM.
-        const nonFunctionalHeroes = this.summaryHeroes?.filter((h) => !functionalMembers.includes(h));
+        const nonFunctionalHeroes = this.heroes?.filter((h) => !functionalMembers.includes(h.userId));
         const hasHeroes = Array.isArray(nonFunctionalHeroes) && nonFunctionalHeroes.length;
         if (hasHeroes) {
+            // use first hero which exists
+            for (const hero of nonFunctionalHeroes) {
+                if (!hero.displayName && !hero.avatarUrl) {
+                    // attempt to look up renderable fields from the m.room.member event if it exists
+                    const member = this.getMember(hero.userId);
+                    if (member) {
+                        return member;
+                    }
+                } else {
+                    // use the Hero supplied values for the room member.
+                    // TODO: It's unfortunate that this function, which clearly only cares about the
+                    //       avatar url, returns the entire RoomMember event. We need to fake an event
+                    //       to meet this API shape.
+                    const heroMember = new RoomMember(this.roomId, hero.userId);
+                    // set the display name and avatar url
+                    heroMember.setMembershipEvent(
+                        new MatrixEvent({
+                            // ensure it's unique even if we hit the same millisecond
+                            event_id: "$" + this.roomId + hero.userId + new Date().getTime(),
+                            type: EventType.RoomMember,
+                            state_key: hero.userId,
+                            content: {
+                                displayname: hero.displayName,
+                                avatar_url: hero.avatarUrl,
+                            },
+                        }),
+                    );
+                    return heroMember;
+                }
```

Comment:
this logic feels broken.

If we have a SSSS hero who *hasn't* set a displayname or avatar, then we use `this.getMember()`. Isn't that path meant to be used only for the legacy-hero-path?

Reply:
I asked Kegan and it sounds is intentional if the server doesn't populate either field, it will essentially fall back to the old behaviour:

> If we have a Hero then the server has explicitly sent data back and we should use it. But, in the case of where there is no actual data for this hero (avatar/display name) then rather than send nothing we fallback to the m.room.member event's displayname/avatar_url if they exist, as it might have some data, which is better than displaying nothing. So effectively the codepath for v2 fallback because-there-is-no-Hero and SSS-Hero-has-no-data is the same

I've tweaked the comment a bit to reflect this.

Reply:
I don't really understand, but let's take it to another thread.